File: backend/core/permissions.py
"""DRF permission classes for IHS PBX RBAC.

Permission model
----------------
* Users belong to zero or more :class:`~core.models.Group` rows via
  :class:`~core.models.UserGroup`.
* Each Group can have zero or more :class:`~core.models.GroupPermission`
  rows that name a permission string (e.g. ``'domain_admin'``,
  ``'dialplan_view'``) and carry a boolean ``permission_assigned`` flag.
* A user is considered to *have* a permission when at least one of their
  groups contains a GroupPermission row for that name with
  ``permission_assigned=True``.

All permission checks short-circuit for superusers (``is_superuser=True``).
"""

import logging

from rest_framework.permissions import BasePermission, SAFE_METHODS

from .models import GroupPermission, UserGroup

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Per-user action grants
# ---------------------------------------------------------------------------
# The per-page action grants stored on ``User.allowed_actions`` use frontend
# page keys × canonical actions (view/add/edit/delete). The DRF viewsets, on the
# other hand, gate on FusionPBX-style permission strings (``extension_add``,
# ``ivr_menu_delete``, …). This registry bridges the two so a per-user grant can
# satisfy a viewset's ``action_permissions`` check regardless of the app-specific
# prefix. Only the pages opted into action-level control are listed; permission
# strings not present here are never satisfied by ``allowed_actions`` and fall
# back to group-based checks (unchanged behavior).
#
# Shape: permission_string -> (page_key, action)
ACTION_PERMISSION_REGISTRY = {}


def _register_page_actions(page_key, prefix):
    """Map ``{prefix}_{view,add,edit,delete}`` permission strings to (page, action)."""
    for action in ('view', 'add', 'edit', 'delete'):
        ACTION_PERMISSION_REGISTRY[f'{prefix}_{action}'] = (page_key, action)


# page_key (matches the frontend route/allowed_pages key) -> permission prefix
_PAGE_PERMISSION_PREFIXES = {
    'extensions':    'extension',
    'ring-groups':   'ring_group',
    'ivr-menus':     'ivr_menu',
    'call-flows':    'call_flow',
    'destinations':  'destination',
    'voicemails':    'voicemail',
    'call-centers':  'call_center',
    'conferences':   'conferences',
    'working-hours': 'working_hours',
}
for _page, _prefix in _PAGE_PERMISSION_PREFIXES.items():
    _register_page_actions(_page, _prefix)


def user_action_allowed(user, permission_name: str) -> bool:
    """True if *user*'s per-user ``allowed_actions`` grants ``permission_name``.

    Returns False for permission strings not covered by the action registry, so
    callers can fall through to group-based checks. Never grants for the empty
    string.
    """
    mapping = ACTION_PERMISSION_REGISTRY.get(permission_name)
    if not mapping:
        return False
    page, action = mapping
    grants = getattr(user, 'allowed_actions', None) or {}
    page_actions = grants.get(page) or []
    return action in page_actions


# ---------------------------------------------------------------------------
# Low-level helper
# ---------------------------------------------------------------------------

def check_permission(user, permission_name: str) -> bool:
    """Return ``True`` if *user* holds the named FusionPBX permission.

    The lookup traverses:
        user  ->  UserGroup  ->  Group  ->  GroupPermission(permission_name)

    A superuser always returns ``True`` without hitting the database.

    Parameters
    ----------
    user:
        A :class:`~core.models.User` instance (may be anonymous).
    permission_name:
        The raw FusionPBX permission string, e.g. ``'domain_admin'``.
    """
    if not user or not user.is_authenticated:
        return False
    if getattr(user, 'is_superuser', False):
        return True

    # Per-user action grants (allowed_actions) can satisfy action-level
    # permission strings without group membership. Group permissions below
    # remain a valid alternative source.
    if user_action_allowed(user, permission_name):
        return True

    return GroupPermission.objects.filter(
        group__user_groups__user=user,
        permission_name=permission_name,
        permission_assigned=True,
    ).exists()


def get_user_permissions(user) -> set:
    """Return the full set of permission names held by *user*.

    Returns an empty set for unauthenticated users.
    """
    if not user or not user.is_authenticated:
        return set()
    return set(
        GroupPermission.objects.filter(
            group__user_groups__user=user,
            permission_assigned=True,
        ).values_list('permission_name', flat=True)
    )


def get_user_group_names(user) -> set:
    """Return the set of group names the user belongs to."""
    if not user or not user.is_authenticated:
        return set()
    return set(
        UserGroup.objects.filter(user=user)
        .values_list('group__group_name', flat=True)
    )


# ---------------------------------------------------------------------------
# Permission classes
# ---------------------------------------------------------------------------

class IsSuperAdmin(BasePermission):
    """Grants access only to users with ``is_superuser=True``.

    This maps to FusionPBX's top-level superadmin role that can manage all
    domains and system-wide settings.
    """

    message = 'Superadmin access required.'

    def has_permission(self, request, view):
        return (
            request.user is not None
            and request.user.is_authenticated
            and getattr(request.user, 'is_superuser', False)
        )


class IsTenantAdmin(BasePermission):
    """Grants access to users that hold the ``tenant_admin`` permission
    within their tenant, or to superusers.
    """

    message = 'Tenant admin access required.'
    PERMISSION_NAME = 'tenant_admin'

    def has_permission(self, request, view):
        user = request.user
        if not user or not user.is_authenticated:
            return False
        if getattr(user, 'is_superuser', False):
            return True
        # Tenant-scoped admins are modeled via the ``admin_tenants`` M2M
        # (UserTenantAccess), not group permissions. Recognize them here so the
        # serializer/UI notion of "Admin" matches this gate. Object-level tenant
        # scoping is enforced separately in the affected viewsets.
        if user.admin_tenants.exists():
            return True
        return check_permission(user, self.PERMISSION_NAME)


# Backward-compat alias — existing views using IsDomainAdmin keep working.
IsDomainAdmin = IsTenantAdmin


class HasPBXPermission(BasePermission):
    """Generic permission gate driven by a ``required_permission`` attribute
    on the view.

    Usage on a ViewSet or APIView::

        class MyViewSet(ModelViewSet):
            required_permission = 'dialplan_view'
            permission_classes = [IsAuthenticated, HasPBXPermission]

    If the view does not set ``required_permission`` the check is skipped and
    access is granted to any authenticated user (effectively equivalent to
    ``IsAuthenticated``).

    Per-action overrides
    ~~~~~~~~~~~~~~~~~~~~~
    You can also define ``action_permissions`` on the view to map ViewSet
    actions to specific permission names::

        class MyViewSet(ModelViewSet):
            action_permissions = {
                'list':    'dialplan_view',
                'retrieve': 'dialplan_view',
                'create':  'dialplan_edit',
                'update':  'dialplan_edit',
                'partial_update': 'dialplan_edit',
                'destroy': 'dialplan_delete',
            }
            permission_classes = [IsAuthenticated, HasPBXPermission]
    """

    message = 'You do not have the required permission.'

    def has_permission(self, request, view):
        user = request.user
        if not user or not user.is_authenticated:
            return False
        # Superusers and staff admins bypass action-level gating: per-user action
        # grants only constrain standard (non-staff) users, mirroring how
        # allowed_pages works. Admins retain full access to every action.
        if getattr(user, 'is_superuser', False) or getattr(user, 'is_staff', False):
            return True

        # Action-level map takes priority (ViewSet actions). The map may live on
        # the permission subclass (e.g. ExtensionPermission.action_permissions)
        # or be overridden per-view; prefer the view, then this instance.
        action_permissions = (
            getattr(view, 'action_permissions', None)
            or getattr(self, 'action_permissions', None)
            or {}
        )
        action = getattr(view, 'action', None)
        if action and action in action_permissions:
            required = action_permissions[action]
            if required == '':
                return True
            return check_permission(user, required)

        # Fall back to a single required_permission (view first, then instance).
        required = (
            getattr(view, 'required_permission', None)
            if getattr(view, 'required_permission', None) is not None
            else getattr(self, 'required_permission', None)
        )
        if required is None:
            return True  # No restriction defined — allow authenticated users.
        if required == '':
            return True
        return check_permission(user, required)

    def has_object_permission(self, request, view, obj):
        return self.has_permission(request, view)


class IsOwnerOrAdmin(BasePermission):
    """Grants access when the requesting user *owns* the object or is a
    tenant admin / superuser.

    Ownership is determined by comparing ``obj.tenant`` with the user's own
    tenant.  Additionally, if the object has a ``user`` attribute the
    requesting user must match that user (or be an admin).

    Attach after authentication classes::

        permission_classes = [IsAuthenticated, IsOwnerOrAdmin]
    """

    message = 'You do not have permission to access this object.'

    def has_permission(self, request, view):
        return request.user is not None and request.user.is_authenticated

    def has_object_permission(self, request, view, obj):
        user = request.user
        if not user or not user.is_authenticated:
            return False
        if getattr(user, 'is_superuser', False):
            return True
        if check_permission(user, 'tenant_admin'):
            return True

        # Tenant ownership check.
        obj_tenant = getattr(obj, 'tenant_id', None)
        user_tenant = getattr(user, 'tenant_id', None)
        if obj_tenant is not None and user_tenant is not None:
            if obj_tenant != user_tenant:
                return False

        # Direct user ownership check (e.g. UserSetting, UserLog).
        obj_user = getattr(obj, 'user_id', None) or getattr(obj, 'user', None)
        if obj_user is not None:
            # Compare primary keys to avoid object identity issues.
            obj_user_pk = obj_user.pk if hasattr(obj_user, 'pk') else obj_user
            return obj_user_pk == user.pk

        return True


class IsSameTenantOrAdmin(BasePermission):
    """Restricts object access to the current request tenant.

    Used on ViewSets where all objects must belong to ``request.tenant``
    unless the caller is a superuser or tenant admin.
    """

    message = 'Cross-tenant access is not permitted.'

    def has_permission(self, request, view):
        return request.user is not None and request.user.is_authenticated

    def has_object_permission(self, request, view, obj):
        user = request.user
        if not user or not user.is_authenticated:
            return False
        if getattr(user, 'is_superuser', False):
            return True

        if check_permission(user, 'tenant_admin'):
            # Tenant admins restricted to their own tenant.
            user_tenant = getattr(user, 'tenant_id', None)
            obj_tenant = getattr(obj, 'tenant_id', None)
            return obj_tenant is None or obj_tenant == user_tenant

        request_tenant = getattr(request, 'tenant', None)
        request_tenant_id = request_tenant.pk if request_tenant else None
        obj_tenant_id = getattr(obj, 'tenant_id', None)
        return obj_tenant_id is None or obj_tenant_id == request_tenant_id


# Backward-compat alias.
IsSameDomainOrAdmin = IsSameTenantOrAdmin


class ReadOnly(BasePermission):
    """Allows GET, HEAD, OPTIONS requests only.

    Intended for composition with ``|`` in DRF permission classes list.
    """

    def has_permission(self, request, view):
        return request.method in SAFE_METHODS


class IsStaffOrReadOnly(BasePermission):
    """Read for any authenticated user; write only for staff/superusers.

    Mirrors the frontend role tiers (superuser/admin = ``is_staff``/``is_superuser``;
    standard ``user`` has neither). Standard users get a read-only view; only
    admins and superusers may create, update, or delete.
    """

    def has_permission(self, request, view):
        user = request.user
        if not user or not user.is_authenticated:
            return False
        if request.method in SAFE_METHODS:
            return True
        return bool(getattr(user, 'is_staff', False) or getattr(user, 'is_superuser', False))

File: frontend/src/lib/permissions.js
// ── Page-based permissions ────────────────────────────────────────────────────
// Single source of truth for what each role may see/do.
// Role tiers (highest → lowest): superuser ⊃ admin ⊃ user.

export function roleOf(user) {
  if (user?.is_superuser) return 'superuser'
  if (user?.is_staff) return 'admin'
  return 'user'
}

// True if a user of `role` is allowed an item tagged with `requiredRole`.
// Untagged items (requiredRole falsy) are visible to everyone.
export function hasRole(role, requiredRole) {
  if (!requiredRole) return true
  if (requiredRole === 'superuser') return role === 'superuser'
  if (requiredRole === 'admin') return role === 'superuser' || role === 'admin'
  return true
}

// Minimum role required to access a given route path.
// Paths not listed are open to any authenticated user.
// Keep this in sync with the Sidebar nav `role` tags.
export const PAGE_ROLES = {
  'tenants':            'admin',
  'audit-log':          'admin',
  'super-users':        'superuser',
  'tenant-list':        'superuser',
  'domains':            'superuser',
  'gateways':           'superuser',
  'outbound-routes':    'superuser',
  'firewall':           'superuser',
  'global-active-calls':'superuser',
  'freeswitch':         'superuser',
  'freeswitch-log':     'superuser',
  'api-keys':           'superuser',
  'system-log':         'superuser',
  'admin-cdr':          'superuser',
  'admin-inventory':    'superuser',
}

// Which UC/PBX user roles a given role is allowed to create.
// superuser → can create admins + users; admin → users only.
export function creatableRoles(role) {
  if (role === 'superuser') return ['superuser', 'admin', 'user']
  if (role === 'admin') return ['user']
  return []
}

// ── Per-user page grants ───────────────────────────────────────────────────────
// Standard ('user') accounts only see pages explicitly granted to them via the
// user's `allowed_pages` list. Admins/superusers ignore this list and get full
// role-based access.
//
// GRANTABLE_PAGES is the catalog an admin can assign from — it is the set of
// pages NOT gated behind an admin/superuser role (i.e. not in PAGE_ROLES). The
// Dashboard (stored as 'dashboard', route '/') is grantable like any other page.
// Each entry's `path` is what gets stored in allowed_pages. Keep labels/groups in
// sync with the Sidebar.
export const GRANTABLE_PAGES = [
  { group: 'General', items: [
    { path: 'dashboard',           label: 'Dashboard' },
  ]},
  { group: 'Call Management', items: [
    { path: 'extensions',          label: 'Extensions' },
    { path: 'ring-groups',         label: 'Ring Groups' },
    { path: 'ivr-menus',           label: 'IVR Menus' },
    { path: 'call-flows',          label: 'Call Flows' },
    { path: 'destinations',        label: 'DIDs' },
    { path: 'custom-destinations', label: 'Custom Destinations' },
    { path: 'working-hours',       label: 'Working Hours' },
    { path: 'call-centers',        label: 'Call Centers' },
  ]},
  { group: 'Communication', items: [
    { path: 'voicemails',          label: 'Voicemails' },
    { path: 'voicemail-inbox',     label: 'Voicemail Inbox' },
    { path: 'conferences',         label: 'Conferences' },
    { path: 'call-parking',        label: 'Call Parking' },
    { path: 'fax',                 label: 'Fax' },
  ]},
  { group: 'System', items: [
    { path: 'media-files',         label: 'Media Files' },
    { path: 'dialplans',           label: 'Dialplans' },
    { path: 'devices',             label: 'Devices' },
  ]},
  { group: 'Reports', items: [
    { path: 'cdr',                 label: 'Call Detail Records' },
    { path: 'call-recordings',     label: 'Call Recordings' },
    { path: 'stats-report',        label: 'Stats Report' },
  ]},
  { group: 'Monitoring', items: [
    { path: 'active-calls',        label: 'Active Calls' },
    { path: 'registrations',       label: 'Peer Status' },
  ]},
]

// Flat list of all grantable page paths.
export const GRANTABLE_PATHS = GRANTABLE_PAGES.flatMap(g => g.items.map(i => i.path))

// ── Per-page action grants ─────────────────────────────────────────────────────
// Pages opted into action-level (view/add/edit/delete) control. Only these pages
// expose action checkboxes in the user editor and are enforced per-action on the
// backend (see core.permissions._PAGE_PERMISSION_PREFIXES). A page not listed
// here keeps page-level-only behavior: access to the page implies all actions.
// Keys must match the page path stored in allowed_pages.
export const ACTIONS = ['view', 'add', 'edit', 'delete']

export const ACTION_LABELS = {
  view:   'View',
  add:    'Create',
  edit:   'Edit',
  delete: 'Delete',
}

export const ACTION_CONTROLLED_PAGES = [
  'extensions',
  'ring-groups',
  'ivr-menus',
  'call-flows',
  'destinations',
  'voicemails',
  'call-centers',
  'conferences',
  'working-hours',
]

// True if `user` may perform `action` ('view'|'add'|'edit'|'delete') on `page`.
// - Superusers and admins (any is_staff) bypass — full access, like allowed_pages.
// - For pages NOT under action control, access to the page implies all actions.
// - For standard users on an action-controlled page, the action must be present
//   in allowed_actions[page].
export function canPerformAction(user, page, action) {
  if (roleOf(user) !== 'user') return true
  if (!ACTION_CONTROLLED_PAGES.includes(page)) return true
  const grants = user?.allowed_actions && typeof user.allowed_actions === 'object'
    ? user.allowed_actions
    : {}
  const allowed = Array.isArray(grants[page]) ? grants[page] : []
  return allowed.includes(action)
}

// True if `user` may access the route identified by `page` (path without slash).
// - Role-gated pages (in PAGE_ROLES) use the role tier check.
// - For standard users, other pages require an explicit grant in allowed_pages.
// - Admins/superusers always pass non-role-gated pages.
export function canAccessPage(user, page) {
  const role = roleOf(user)
  const required = PAGE_ROLES[page]
  if (required) return hasRole(role, required)
  if (role !== 'user') return true
  // Standard user: only granted pages. The Dashboard ('dashboard') is gated like
  // any other page — it must be explicitly granted.
  const granted = Array.isArray(user?.allowed_pages) ? user.allowed_pages : []
  return granted.includes(page)
}

// Resolve a route path (e.g. '/', '/fax') to its page key used in allowed_pages.
// The root route maps to the 'dashboard' grant key.
export function pageKeyOf(routePath) {
  const stripped = routePath.replace(/^\//, '')
  return stripped === '' ? 'dashboard' : stripped
}

// The path a user should land on after login or when bounced from a denied page.
// Prefers the Dashboard when granted, otherwise the first page the user can see.
// Falls back to '/' (which itself may be denied, but avoids an undefined target).
export function landingPath(user) {
  if (roleOf(user) !== 'user') return '/'
  if (canAccessPage(user, 'dashboard')) return '/'
  const granted = Array.isArray(user?.allowed_pages) ? user.allowed_pages : []
  const first = GRANTABLE_PATHS.find(p => p !== 'dashboard' && granted.includes(p))
  return first ? `/${first}` : '/'
}
